File: tools/commands/ci/fuzz/enforce-result.rs
#!/usr/bin/env rust-script
//! ```cargo
//! [package]
//! edition = "2024"
//! ```

use std::{env, process::ExitCode};

fn main() -> ExitCode {
    let args = env::args().skip(1).collect::<Vec<_>>();
    let [event_name, target, outcome] = args.as_slice() else {
        eprintln!(
            "Usage: rust-script tools/commands/ci/fuzz/enforce-result.rs <event-name> <target> <outcome>"
        );
        return ExitCode::from(1);
    };

    if !matches!(
        event_name.as_str(),
        "pull_request" | "schedule" | "workflow_dispatch"
    ) {
        eprintln!("Unsupported fuzz event: {}", empty_label(event_name));
        return ExitCode::from(1);
    }
    if !matches!(
        outcome.as_str(),
        "success" | "failure" | "cancelled" | "skipped"
    ) {
        eprintln!("Unsupported fuzz outcome: {}", empty_label(outcome));
        return ExitCode::from(1);
    }
    if outcome == "success" {
        println!("Fuzz target {target} completed successfully.");
        return ExitCode::SUCCESS;
    }

    let message = format!("Fuzz target {target} finished with {outcome} on {event_name}.");
    if matches!(event_name.as_str(), "schedule" | "workflow_dispatch") {
        eprintln!("::error::{message} Release evidence must be green.");
        ExitCode::from(1)
    } else {
        eprintln!("::warning::{message} Pull-request fuzzing is advisory.");
        ExitCode::SUCCESS
    }
}

fn empty_label(value: &str) -> &str {
    if value.is_empty() { "empty" } else { value }
}

File: tools/support/compat/fuzz/enforce-result.mjs
import path from "node:path";
import { fileURLToPath } from "node:url";

const outcomes = new Set(["success", "failure", "cancelled", "skipped"]);
const events = new Set(["pull_request", "schedule", "workflow_dispatch"]);
const releaseEvents = new Set(["schedule", "workflow_dispatch"]);

export function fuzzResultPolicy(eventName, outcome) {
  if (!events.has(eventName)) {
    throw new Error(`Unsupported fuzz event: ${eventName || "empty"}`);
  }
  if (!outcomes.has(outcome)) {
    throw new Error(`Unsupported fuzz outcome: ${outcome || "empty"}`);
  }
  const unsuccessful = outcome !== "success";
  return {
    unsuccessful,
    releaseBlocking: unsuccessful && releaseEvents.has(eventName),
  };
}

export function reportFuzzResult(eventName, target, outcome) {
  const policy = fuzzResultPolicy(eventName, outcome);
  if (!policy.unsuccessful) {
    console.log(`Fuzz target ${target} completed successfully.`);
    return 0;
  }

  const message = `Fuzz target ${target} finished with ${outcome} on ${eventName}.`;
  if (policy.releaseBlocking) {
    console.error(`::error::${message} Release evidence must be green.`);
    return 1;
  }
  console.warn(`::warning::${message} Pull-request fuzzing is advisory.`);
  return 0;
}

const entrypoint = process.argv[1]
  ? fileURLToPath(import.meta.url) === path.resolve(process.argv[1])
  : false;
if (entrypoint) {
  const [eventName, target, outcome, extra] = process.argv.slice(2);
  try {
    if (!eventName || !target || !outcome || extra != null) {
      throw new Error(
        "Usage: rust-script tools/commands/ci/fuzz/enforce-result.rs <event-name> <target> <outcome>",
      );
    }
    process.exitCode = reportFuzzResult(eventName, target, outcome);
  } catch (error) {
    console.error(error instanceof Error ? error.message : String(error));
    process.exitCode = 1;
  }
}
